Add InternalDuration class for Temporal support

catamorphism opened the pull request:
https://bugs.webkit.org/show_bug.cgi?id=223166

Reviewed by NOBODY (OOPS!).

Add a new InternalDuration class for use by Temporal,
following the Temporal spec that uses a separate Duration and
InternalDuration type. In a future PR, code will be modified
to convert Durations to InternalDurations for internal computation,
then convert back when necessary. This is necessary to maintain
precision.

* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::dateDurationSign):
(JSC::ISO8601::ISO8601::InternalDuration::sign const):
(JSC::ISO8601::InternalDuration::combineDateAndTimeDuration):
* Source/JavaScriptCore/runtime/ISO8601.h:<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9827fa88b04428c7e707de4d3557804c3085c61b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103182 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22857 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108350 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53825 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23192 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31357 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78436 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35374 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106188 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17983 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93095 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58769 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11137 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53180 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95857 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87636 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11200 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110727 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101793 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30319 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22355 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87425 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30684 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89294 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87056 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31915 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9628 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24643 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30246 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35568 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125426 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30055 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34792 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33381 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31617 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/ISO8601.h:
```diff
@@ -170,6 +170,43 @@ class ExactTime {
+        : m_dateDuration(Duration()), m_time(0) { }
+    static constexpr Int128 maxTimeDuration = 9'007'199'254'740'992 * ExactTime::nsPerSecond - 1;
+
+    int32_t sign() const;
```
Should we implement this `sign` method? Maybe https://tc39.es/proposal-temporal/#sec-durationsign ?

catamorphism replied:
Done in [656e2e3](https://github.com/WebKit/WebKit/pull/43980/commits/656e2e3542816ef9c2ebdf44df642c6fa2d7d842) (it was already implemented, I just hadn't pulled it in when I was splitting up the original PR).

sosukesuzuki commented on Source/JavaScriptCore/runtime/ISO8601.h:
```diff
@@ -170,6 +170,43 @@ class ExactTime {
+   to maxTimeDuration, where
+   maxTimeDuration = 2**53 × 10**9 - 1 = 9,007,199,254,740,991,999,999,999.
+   It represents the portion of a Temporal.Duration object that deals with time
+   units, but as a combined value of total nanoseconds.  */
```
I think it would be better to use line style comment (`// foo`) here for coding style consistency.

catamorphism replied:
Done in [aed29bd](https://github.com/WebKit/WebKit/pull/43980/commits/aed29bd52829c0a813d98fdca0332fab2e82e87d)

ptomato commented on Source/JavaScriptCore/runtime/ISO8601.cpp:
```diff
@@ -1599,6 +1599,52 @@ Int128 ExactTime::round(Int128 quantity, unsigned increment, TemporalUnit unit,
+    if (dateSign && timeSign && dateSign != timeSign) {
+        throwRangeError(globalObject, scope, "Date and time have different signs"_s);
+        return { };
+    }
```
In the spec text, this is an assertion. (I think I remember changing this at some point, maybe around last October.) Is it feasible to make it an assertion here as well? In the spec, there's only one caller of this operation where we could end up with mismatched signs, and it is checked explicitly before calling CombineDateAndTimeDuration. At all other call sites, the signs can be trivially deduced to be either equal or zero.

This would allow removing the `globalObject` parameter.

catamorphism replied:
Fixed in [a86202d](https://github.com/WebKit/WebKit/pull/43980/commits/a86202d7ed692fdb479134c1ac06cfa9399a893f) (the only reason I didn't do it that way in the first place is that I wrote this code before the spec was changed, most likely).

catamorphism committed:
Add InternalDuration class for Temporal support
https://bugs.webkit.org/show_bug.cgi?id=223166

Reviewed by NOBODY (OOPS!).

Add a new InternalDuration class for use by Temporal,
following the Temporal spec that uses a separate Duration and
InternalDuration type. In a future PR, code will be modified
to convert Durations to InternalDurations for internal computation,
then convert back when necessary. This is necessary to maintain
precision.

* Source/JavaScriptCore/runtime/ISO8601.cpp:
(JSC::ISO8601::dateDurationSign):
(JSC::ISO8601::ISO8601::InternalDuration::sign const):
(JSC::ISO8601::InternalDuration::combineDateAndTimeDuration):
* Source/JavaScriptCore/runtime/ISO8601.h:
